### DISCUSSION TITLE
env secret undefined in preview

### DISCUSSION THREAD

**Discussion by harm-meijer** [vercel/vercel#4145] [General]
In my code I log the env that is connected to the secret:

    console.log('key is:', process.env.VUE_APP_GOOGLE_MAPS_API_KEY);

In now.json I connect the value:

    {
      "env": {
        "VUE_APP_GOOGLE_MAPS_API_KEY": "@google-maps-api-key-preview-w7nc"
      },
    }

I created `google-maps-api-key-preview-w7nc` on vercel.com but creating it with cli had it logging undefined as well: `now secrets add sunrise-spa.prod.google_maps_api_key  the-value`.

Output of `now secrets list` is:

> google-maps-api-key-preview-w7nc            18m ago
> sunrise-spa.prod.google_maps_api_key        2h ago

Preview is here: https://sunrise-spa-wet5iqd1a.now.sh/DE/en/ logging undefined.

**Comment by lucleray** [comment]
Hi @harm-meijer,

That should definitely work! Can you make sure that the environment variable you defined in the dashboard matches `VUE_APP_GOOGLE_MAPS_API_KEY`?

According to the output of `now secrets list` you shared, you created a environment variable named `GOOGLE_MAPS_API_KEY` but no environment variable named `VUE_APP_GOOGLE_MAPS_API_KEY`.

Then, you don't need to add anything in your `now.json` file or use `now secrets add`.

**Reply by harm-meijer** [reply]
Hi @lucleray 
`According to the output of now secrets list you shared, you created a environment variable named GOOGLE_MAPS_API_KEY` I thought I created a secret named `google-maps-api-key` (the last part is what the dashboard added) and try to link it with an environment variable named [VUE_APP_GOOGLE_MAPS_API_KEY](https://github.com/commercetools/sunrise-spa/blob/demo-ready/now.json#L6).

When I log [VUE_APP_GOOGLE_MAPS_API_KEY](https://github.com/commercetools/sunrise-spa/blob/demo-ready/src/main.js#L44) it is [undefined](https://sunrise-spa-wet5iqd1a.now.sh/DE/en/).

Happy to hear it should work, now if only it were the case that it does work. There must be a config missing somewhere but can't figure what or where.

**Reply by lucleray** [reply]
Looking at the code, it seems that you need to provide the environment variable at build time:
```
{
  "build": {
    "env": {
      "VUE_APP_GOOGLE_MAPS_API_KEY": "@google-maps-api-key-preview-w7nc"
    }
  }
}
```

Can you try this and let me know if it solves it for you?

**Reply by lucleray** [reply]
However, I would not advise to use the secret generated for the `google_maps_api_key` environment variable, because if you delete the environment variable, the secret will be deleted as well.

I'd recommend that you add the environment variables through our dashboard instead of using `now.json`. You can add the environment variable with the name `VUE_APP_GOOGLE_MAPS_API_KEY` and the value of your api key. The environment variable will automatically be made available both for Build and Runtime and you can omit it in `now.json`.

**Reply by harm-meijer** [reply]
Thank you for your suggestion. I removed the keys with cli `now secrets remove` and added them with the dashboard. The command: `now secrets list` gives me:

>vue-app-google-maps-api-key-preview-59db     12m ago
vue-app-google-maps-api-key-production-ghc3  13m ago

Then removed the [env part](https://github.com/commercetools/sunrise-spa/commit/317f8e13200a969803ab3c28d986440d1cf02abe) from now.json and pushed the changes. It still [logs undefined](https://sunrise-spa-git-demo-ready.commercetools.now.sh/DE/en/) though.

**Reply by lucleray** [reply]
@harm-meijer The [deployment](https://github.com/commercetools/sunrise-spa/commit/317f8e13200a969803ab3c28d986440d1cf02abe) you shared happened before your removed the env variable named `google_maps_api_key` from your project, so it attempted to use that env variable. Since the secret attached to the variable was removed, the deployment failed.

The deployment attaches Environment Variables that are present in your project settings at the moment you deploy. If you update the Environment Variables, you need to deploy again to use the newly configured ones.

Can you try deploying again to see if it works as expected now?

**Reply by harm-meijer** [reply]
Thank you, I missed that it failed to publish. I it is working now.
